### PR TITLE
Variable Size Kernels

### PR DESCRIPTION
**Pull request by antiochp**
[Rendered link to RFC document](https://github.com/antiochp/grin-rfcs/blob/variable_size_kernels/text/0000-variable-size-kernels.md)

This is a pre-requisite for "relative kernels" as we need to be able to introduce new kernel feature variants, with additional associated data.

The proposal is to support kernels of variable size, with each kernel variant supporting additional data specific to the variant. Height locked kernels have an associated lock height for example.

### CODE REVIEW

**Comment by tromp** [comment]
thx for the write-up, Antioch!
there's a typo in "benfit"; also rather than give byte savings over 1M kernels, iI suggest saying it's about a 7% space savings per plain kernel (105/113 bytes).

**Comment by lehnberg** [comment]
Just took a first pass at this. I actually understood the proposal, which says a lot about the quality of your writing @antiochp - nice job. Reads clear and concise.

A modest suggestion would be to write the RFC as if it's already been accepted and is current documentation, i.e.:

_"The proposal is to have variant specific data follow the feature byte"_ would become something more like _"Variant specific data follow the feature byte"_; and

_"The initial feature byte would determine the number of subsequent bytes to read for variant specific data."_ becomes _"The initial feature byte determines the number..."_

Similarly, protocol version 1 should be referred to as a previous version, and the "proposed" version 2 should be referred to as the actual current version, if that makes sense. So that it's not a proposal, but the status quo.

(These are not exhaustive examples)

Rationale being that if the RFC is rejected, then the text will not exist in any case, but if it gets merged, then it will be in the repo and act as an instruction essentially, until the point it gets  superseded by some other RFC.

**Comment by antiochp** [comment]
@lehnberg That makes a lot of sense - I'll rework as suggested.

**Comment by antiochp** [comment]
Still got one remaining section to complete (txhashset fast sync support). But promoting this to a full PR.

**Comment by lehnberg** [comment]
In accordance with our RFC process, as of today this can be considered in Final Comment Period with a disposition of merge

**Comment by lehnberg** [comment]
The Final Comment Period is due to expire for this RFC tomorrow. The proposal will then be merged and considered accepted unless there is a strong case made against it.

**Comment by jaspervdm** [comment]
This might be a stupid question, but do block headers commit to the full kernels that are contained in the block?
- If yes: does this versioning affect this? I.e. if i convert the kernels inside a block from protocol v1 to v2, will it invalidate the PoW?
- If no: what does it commit to, the excess only? Wouldn't that make kernels malleable to some extent? Maybe this is not a big deal because creating a kernel with the same public excess but different features requires knowledge of the excess, so only the original creators could do it.

**Comment by DavidBurkett** [comment]
> This might be a stupid question, but do block headers commit to the full kernels that are contained in the block?
> 
> * If yes: does this versioning affect this? I.e. if i convert the kernels inside a block from protocol v1 to v2, will it invalidate the PoW?
> * If no: what does it commit to, the excess only? Wouldn't that make kernels malleable to some extent? Maybe this is not a big deal because creating a kernel with the same public excess but different features requires knowledge of the excess, so only the original creators could do it.

No, block headers just commit to the excess, MMR roots, and MMR sizes. Kernels aren't malleable because changing them would change the MMR root

**Comment by antiochp** [comment]
Headers commit to the kernel MMR root. Which in turn commits to the hashes (and positions) of all the kernels contained in the kernel MMR.

So headers don't commit to the full kernels contained in the block directly. 
But they _do_ commit to all kernel hashes in the kernel MMR in its entirety (which includes the kernels in the block in question).

tl;dr not the _full kernels in the block_, but the _kernel hashes in all blocks_

Kernel hashes currently are based on the full kernels (even for 0 fee on coinbase kernels etc). And we need to be careful to preserve this for historical data even when moving to future protocol versions.

File: text/0000-variable-size-kernels.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
+
+- Title: variable-size-kernels
+- Authors: [Antioch Peverell](mailto:apeverell@protonmail.com)
+- Start date : Aug 13, 2019
```
Comment (quentinlesceller) [inline text/0000-variable-size-kernels.md]: Start date : Aug 13, 2019 -> Start date: Aug 13, 2019

File: text/0000-variable-size-kernels.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+# Motivation
+[motivation]: #motivation
+
+We were originally including both fee and lock_height on _every_ kernel in the binary serialization format. We were storing a fee of 0 on coinbase kernels and a lock_height of 0 on both plain and coinbase kernels. Kernels are never pruned and must be maintained forever so this overhead is relatively expensive. By only serializing data strictly necessary for each kernel variant we minimize storage and transmission costs. This also provides the flexibility necessary to introduce new kernel variants in the future, for example the proposed [relative kernels][0].
+
+# Community-level explanation
+[community-level-explanation]: #community-level-explanation
+
+Each transaction kernel variant may have associated data. For example height locked kernels include an associated lock height and non-coinbase kernels have an associated fee. Each kernel variant serializes to a fixed size in bytes but this size may be different for each kernel variants. This allows kernels to be serialized efficiently and provides flexibility to introduce new kernel variants that have additional associated data in the future.
```
Comment (quentinlesceller) [inline text/0000-variable-size-kernels.md]: For example height locked -> For example, height locked (missing comma)

File: text/0000-variable-size-kernels.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+The following p2p messages include serialized transaction kernels -
+
+* transaction
+* full block
+* compact block (incl. coinbase kernel)
+
+Each node has a "local" protocol version. Nodes exchange protocol versions during the initial handshake when setting up a connection.
+
+If both nodes are running protocol version 2 then no translation is required and transaction kernels can be serialized using protocl version 2.
```
Comment (quentinlesceller) [inline text/0000-variable-size-kernels.md]: using protocl version 2. -> using protocol version 2.

File: text/0000-variable-size-kernels.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+Node B will broadcast transactions and blocks using protocol version 1. Node A will need to use the previous protocol version and not the local version when deserializing these messages.
+
+__Sending messages using previous protocol version__
+
+Node A will need to ensure anything broadcast to Node B is compatible with protocol version 1. In both cases node A is responsible for translating to and from the previous protocol version.
+
+#### Local db storage
+
+Each node has a "db" protocol version. All entries in the db serialize/deserialize using that protocol version. Nodes support a process for local migration of data. On startup the db is inspected and if necessary a migration is performed upgrading all entries in the db to the latest protocol version. This process can be disabled locally to allow nodes to run against old databases without upgrading the protocol version. This is useful in cases where nodes do not wish to immediately upgrade the db, allowing for previous versions of code to run without problems.
```
Comment (quentinlesceller) [inline text/0000-variable-size-kernels.md]: On startup the db -> On startup, the db (missing comma)

File: text/0000-variable-size-kernels.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+#### Kernel MMR storage
+
+Each node maintains a kernel MMR as part of the txhashset data structure. Each node has an MMR protocol version. The node will serialize/deserialize kernel entries in the kernel MMR data file according to the MMR protocol version.
+
+##### Fast sync (txhashset.zip state file)
+
+Internally nodes are free to use any protocol version but the kernel MMR is also provided to other nodes during initial fast sync. A node joining the network requests a txhashset state file and this includes the kernel MMR. It is important that nodes send and receive the txhashset file using a compatible protocol version. This is a special case of "p2p messages" above with the txhashset state file provided as an attachment to the txhashset message.
+
+If node A (protocol version 2) requests a txhashset from node B (protocol version 1) then it must read the file using protocol version 1. Similarly if node B requests txhashset from node A then node A must provide the file such that it is compatible with protocol version 1.
```
Comment (quentinlesceller) [inline text/0000-variable-size-kernels.md]: Similarly if node B requests -> Similarly, if node B requests (missing comma)

File: text/0000-variable-size-kernels.md [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+	"excess_sig": "3311...b969"
+  },
+  ...
+]
+```
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+Support for variable size kernels allows new kernel variants to be introduced in the future. One concrete exampe of this is [relative kernels][0]. Kernels with relative lock heights will have an associated reference to a prior kernel in addition to a lock height based on that prior kernel.
```
Comment (quentinlesceller) [inline text/0000-variable-size-kernels.md]: One concrete exampe -> One concrete example

**Comment by lehnberg** [comment]
🎉 **Wohooo! This RFC has now been merged!** 🤸‍♀️
Tracking issue: mimblewimble/grin#3038
